### PR TITLE
[1/2] Add AWS IAM Authenticator hooks

### PR DESCRIPTION
**Pull request by chrisdoherty4**
### Description

This change consolidates AWS IAM Authentication installation to the `awsiamauth` package enabling new workflow hooks to be written and registered with the new create management cluster workflow.

The existing workflow code paths are honored as they call the same APIs created for new workflow hooks.

The AWS IAM Auth implementation hasn't been touched so should be functionally equivalent. A separate PR will address exported APIs and unit tests against the existing code in the `awsiamauth` package.

### Testing

The `local-e2e` make target was run with a Docker AWS IAM Auth test. Additional manual verification was performed by using the AWS Kubeconfig to retrieve pods from the Docker cluster.

### CODE REVIEW

File: pkg/awsiamauth/hook.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/workflow"
+	"github.com/aws/eks-anywhere/pkg/workflow/management"
+	"github.com/aws/eks-anywhere/pkg/workflow/workflowcontext"
+)
+
+type HookRegistrar struct {
+	*AwsIamAuth
+	Spec *cluster.Spec
```
Comment (g-gaston) [inline pkg/awsiamauth/hook.go:15]: For structs with tons of members I can buy that it makes the code a bit cleaner, but for this I would vote to have a constructor and make the members private.

It makes clear what is mandatory and what's not. Also, it stops anyone from tinkering with the inner members directly once the struct is constructed.

File: pkg/awsiamauth/hook.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package awsiamauth
```
Comment (g-gaston) [inline pkg/awsiamauth/hook.go:1]: What do you think about moving this to `pkg/awsiamauth/hook/`?
It avoids importing the workload packages from the main business logic packages.

It's not necessary with this one in particular but I feel it sets a good precedence for the future.

File: pkg/dependencies/factory.go [hunk 3]
```diff
@@ -667,7 +667,7 @@ func (f *Factory) WithAwsIamAuth() *Factory {
 		}
 		certgen := crypto.NewCertificateGenerator()
 		clusterId := uuid.New()
-		f.dependencies.AwsIamAuth = awsiamauth.NewAwsIamAuth(certgen, clusterId)
```
Comment (g-gaston) [inline pkg/dependencies/factory.go:670]: I think we need to add this at the top of `WithAwsIamAuth`

```go
f.WithKubectl().WithWriter()
```

**Review by g-gaston — COMMENTED** [review]
This is great

File: pkg/awsiamauth/hook.go (hunk 2 above)
Comment (chrisdoherty4) [reply pkg/awsiamauth/hook.go:1]: I don't think its necessary in this case. Equally, I could see it being appropriate in other cases.

In general, I think its good to stick to the rule "package A provides everything to do with A" and only break it down if needed.

File: pkg/awsiamauth/hook.go (hunk 1 above)
Comment (chrisdoherty4) [reply pkg/awsiamauth/hook.go:15]: For `Spec` in particular I'm not sure why I did it this way - will adjust. 

The `AwsIamAuth` is embedded so will have to remain exposed. That's done intentionally for convenience so consumers can use the registrar as anything satisfying existing AWS IAM interfaces.

Comment (chrisdoherty4) [reply pkg/awsiamauth/hook.go]: Done

File: pkg/dependencies/factory.go (hunk 3 above)
Comment (chrisdoherty4) [reply pkg/dependencies/factory.go:687]: Done

File: pkg/awsiamauth/hook.go (hunk 2 above)
Comment (chrisdoherty4) [reply pkg/awsiamauth/hook.go:1]: We agreed on my response above.

**Pull request merged**